test/fw/ptl/utils/pbs_snaputils.py

```diff
@@ -0,0 +1,1777 @@
+                                   self.snapdir + "? (y/n) ")
+                if answer == "y":
+                    shutil.rmtree(self.snapdir)
+                else:
+                    raise UserQuit
+
+        os.mkdir(self.snapdir)
+
+        if self.create_tar is True:
+            self.outtar_fd = tarfile.open(self.outtar_path, "w:gz")
+
+        # Create the following directories: server, job, node, comm, queue,
+        # hook, scheduler, reservation, resource, datastore, pbs, core_file_bt
+        # and system
+        dirs_in_snapshot = ['server', 'job', 'node', 'hook', 'scheduler',
+                            'reservation', 'datastore', 'pbs', 'core_file_bt',
+                            'system', SCHED_PRIV_PATH, SVR_PRIV_PATH,
+                            MOM_PRIV_PATH, ACCT_LOGS_PATH, COMM_LOGS_PATH,
+                            SVR_LOGS_PATH, SCHED_LOGS_PATH, MOM_LOGS_PATH]
+        for item in dirs_in_snapshot:
+            rel_path = os.path.join(self.snapdir, item)
+            os.makedirs(rel_path, 0755)
+
+    def __grep_file(self, fd_in, fd_out, pattern):
+        """
+        Grep the file given for the pattern specified
+
+        :param fd_in: file object of the file to grep
+        :type fd_in: file
+        :param fd_out: file object of the file to store grep'd contents
+        :type fd_out: file
+        :param pattern: the pattern to grep for
+        :type pattern: str
+        """
+        for line in fd_in:
+            match = re.search(pattern, line)
+            if match:
+                fd_out.write(line)
+
+    def __capture_cmd_output(self, host, out_path, cmd, grep_pbs=False,
+                             skip_anon=False):
+        """
+        Run a command on the host specified and capture its output
+
+        :param host: the host to run the command on
+        :type host: str
+        :param out_path: path of the output file for this command
+        :type out_path: str
+        :param cmd: The command to execute
+        :type cmd: list
+        :param grep_pbs: Should do "grep pbs" on the output of the command?
+        :type grep_pbs: bool
+        :param skip_anon: Skip anonymization even though anonymize is True?
+        :type skip_anon: bool
+        """
+        self.logger.debug("capturing " + str(cmd))
+
+        if "qmgr" in cmd[0]:
+            # qmgr -c is being called
+            if self.du.is_localhost(host) is False:
+                # For remote hosts, wrap qmgr's command in quotes
+                cmd[2] = "\'" + cmd[2] + "\'"
+
+        with open(out_path, "w") as out_fd:
+            try:
+                cmd_stdout = out_fd
+                if grep_pbs is True:
+                    # Set stdout to a temp file
+                    fd, fn = self.du.mkstemp()
+                    tempfd = os.fdopen(fd, "w+")
+                    cmd_stdout = tempfd
+                self.du.run_cmd(host, cmd=cmd, stdout=cmd_stdout,
```

Comment:
i didn't know that run command could do this, thanks for pointing this out, it makes things much easier!